tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,143 @@
+
+    config = {"ca-common-name": "CN_CA"}
+    await ops_test.model.deploy(TLS_CERTIFICATES_APP_NAME, channel="stable", config=config)
+
+    await ops_test.model.wait_for_idle(
+        apps=[TLS_CERTIFICATES_APP_NAME], status="active", timeout=1000
+    )
+
+    # Relate it to OpenSearch to set up TLS.
+    await ops_test.model.relate(OPENSEARCH_APP_NAME, TLS_CERTIFICATES_APP_NAME)
+    await ops_test.model.wait_for_idle(
+        apps=[OPENSEARCH_APP_NAME, TLS_CERTIFICATES_APP_NAME], status="active", timeout=1000
+    )
+
+    async with ops_test.fast_forward():
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications[APP_NAME].units) == NUM_UNITS_APP
+        )
+        await ops_test.model.wait_for_idle(apps=[APP_NAME], timeout=1000, idle_period=30)
+
+    assert ops_test.model.applications[APP_NAME].status == "blocked"
+
+    pytest.relation = await ops_test.model.relate(OPENSEARCH_APP_NAME, APP_NAME)
+    await ops_test.model.wait_for_idle(
+        apps=[OPENSEARCH_APP_NAME, APP_NAME], status="active", timeout=1000
+    )
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_in_place_upgrade_http(ops_test: OpsTest):
+    leader_unit = None
+    for unit in ops_test.model.applications[APP_NAME].units:
+        if await unit.is_leader_from_status():
+            leader_unit = unit
+    assert leader_unit
+
+    action = await leader_unit.run_action("pre-upgrade-check")
+    await action.wait()
+
+    # ensure action completes
+    time.sleep(10)
+
+    # ensuring app is safe to upgrade
+    assert "upgrade-stack" in get_relation_data(
+        model_full_name=ops_test.model_full_name, unit=f"{APP_NAME}/0", endpoint="upgrade"
+    )
+
+    await ops_test.model.applications[APP_NAME].refresh(path=pytest.charm)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", timeout=1000, idle_period=120
+    )
+
+    assert await access_all_dashboards(ops_test)
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_switch_tls_on(ops_test: OpsTest):
+    """Test HTTPS access to each dashboard unit."""
+    # Relate it to OpenSearch to set up TLS.
+    await ops_test.model.relate(APP_NAME, TLS_CERTIFICATES_APP_NAME)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME, TLS_CERTIFICATES_APP_NAME], status="active", timeout=1000
+    )
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_in_place_upgrade_https(ops_test: OpsTest):
+    leader_unit = None
+    for unit in ops_test.model.applications[APP_NAME].units:
+        if await unit.is_leader_from_status():
+            leader_unit = unit
+    assert leader_unit
+
+    action = await leader_unit.run_action("pre-upgrade-check")
+    await action.wait()
+
+    # ensure action completes
+    time.sleep(10)
+
+    # ensuring app is safe to upgrade
+    assert "upgrade-stack" in get_relation_data(
```

Comment:
same as above - can we add a check on the content of the stack?

Reply:
Done.